### PR TITLE
fix(spans): Support lowercase redis commands

### LINKED ISSUE #3643: Parse & scrub lowercase redis commands
Not all redis integrations send uppercase redis commands, which results in redis spans being scrubbed as `*`. See

https://github.com/getsentry/relay/blob/c6d911004dc743289c5142ac0a54e6df17ac1a42/relay-event-normalization/src/normalize/span/description/mod.rs#L1142-L1143

Solution: Either use a proper redis parser or allowlist some of the most common commands (`get`, `set`, `evalsha`, ...).

### PR DESCRIPTION
**Pull request by jjbayer**
Not all redis integrations send uppercase redis commands, which results in redis spans being scrubbed as *. See

https://github.com/getsentry/relay/blob/c6d911004dc743289c5142ac0a54e6df17ac1a42/relay-event-normalization/src/normalize/span/description/mod.rs#L1142-L1143

This PR removes the redis regex and introduces a list of redis command that can be prefix-searched.

Fixes https://github.com/getsentry/relay/issues/3643

### CODE REVIEW

File: relay-event-normalization/src/normalize/span/description/redis.rs [hunk 1]
```diff
@@ -0,0 +1,591 @@
...
+        })
+        .ok()?;
+    Some(REDIS_COMMANDS[index])
+}
+
+/// List of redis commands.
+///
+/// See <https://redis.io/docs/latest/commands/>.
+const REDIS_COMMANDS: &[&str] = &[
+    "ACL",
```
Comment (Dav1dde) [inline relay-event-normalization/src/normalize/span/description/redis.rs:18]: This assertion fails:

```rs
        assert_eq!(matching_redis_command("ACL CAT"), Some("ACL CAT"));
```

So does this one:
```rs
        assert_eq!(matching_redis_command("ACL cattington"), Some("ACL"));
```

File: relay-event-normalization/src/normalize/span/description/redis.rs [hunk 2]
```diff
@@ -0,0 +1,591 @@
+use std::cmp::Ordering;
+
+/// Returns a redis command if it is a case-insensitive prefix of `seek`.
+pub fn matching_redis_command(seek: &str) -> Option<&str> {
+    let seek = seek.as_bytes();
+    let index = REDIS_COMMANDS
+        .binary_search_by(|probe| {
+            let probe = probe.as_bytes();
+            // step 1 - compare prefixes
```
Comment (Dav1dde) [inline relay-event-normalization/src/normalize/span/description/redis.rs]: It would be great if you can add a few more comments here and some basic examples for the branches/match arms, this is quite hard to think through without.

Same for the second check.

**Comment by Dav1dde** [comment]
Also might be fun to look into [`aho-corasick`](https://docs.rs/aho-corasick/), it has [`MatchKind::LeftmostLongest`](https://docs.rs/aho-corasick/latest/aho_corasick/enum.MatchKind.html) and [`ascii_case_insensitive`](https://docs.rs/aho-corasick/latest/aho_corasick/struct.AhoCorasickBuilder.html#method.ascii_case_insensitive).

File: relay-event-normalization/src/normalize/span/description/redis.rs (hunk 1 above)
Comment (jjbayer) [reply relay-event-normalization/src/normalize/span/description/redis.rs:18]: Whoops good catch, that's what you get for fooling around with low level stuff.

**Comment by jjbayer** [comment]
> Also might be fun to look into [`aho-corasick`](https://docs.rs/aho-corasick/), it has [`MatchKind::LeftmostLongest`](https://docs.rs/aho-corasick/latest/aho_corasick/enum.MatchKind.html) and [`ascii_case_insensitive`](https://docs.rs/aho-corasick/latest/aho_corasick/struct.AhoCorasickBuilder.html#method.ascii_case_insensitive).

I also considered using a [trie](https://en.wikipedia.org/wiki/Trie) for this, e.g. https://docs.rs/prefix-trie/latest/prefix_trie/.

**Comment by loewenheim** [comment]
> Also might be fun to look into [`aho-corasick`](https://docs.rs/aho-corasick/), it has [`MatchKind::LeftmostLongest`](https://docs.rs/aho-corasick/latest/aho_corasick/enum.MatchKind.html) and [`ascii_case_insensitive`](https://docs.rs/aho-corasick/latest/aho_corasick/struct.AhoCorasickBuilder.html#method.ascii_case_insensitive).

… and everything already depends on it anyway, right?

**Comment by Dav1dde** [comment]
> I also considered using a [trie](https://en.wikipedia.org/wiki/Trie) ...

Yes! Trie's are very cool data structures.

> … and everything already depends on it anyway, right?

yep^^

**Comment by Dav1dde** [comment]
But tbf, nothing wrong with a good old binary search!

File: relay-event-normalization/src/normalize/span/description/redis.rs [hunk 3]
```diff
@@ -0,0 +1,610 @@
...
+        }
+    }
+    longest_match
+}
+
+fn is_ascii_prefix_lowercase(a: &str, b: &str) -> bool {
+    (a.len() <= b.len())
+        && a.chars()
+            .zip(b.chars())
+            .all(|(a, b)| a.to_ascii_lowercase() == b.to_ascii_lowercase())
```
Comment (Dav1dde) [inline relay-event-normalization/src/normalize/span/description/redis.rs]: `a.len() <= b.len() && let Some(b) = b.get(..a.len()) && a.eq_ascii_ignore_case(b)`

Can also just compare bytes if you're checking ascii only.

File: relay-event-normalization/src/normalize/span/description/redis.rs [hunk 4]
```diff
@@ -0,0 +1,610 @@
+/// Returns a redis command if it is a case-insensitive prefix of `seek`.
+pub fn matching_redis_command(needle: &str) -> Option<&str> {
+    let mut longest_match = None;
+    for command in REDIS_COMMANDS {
```
Comment (Dav1dde) [inline relay-event-normalization/src/normalize/span/description/redis.rs]: Seems quite expensive, could also build a simple Regex in a `LazyLock` (or aho-corasick, which the regex probably compiles to).

File: relay-event-normalization/src/normalize/span/description/redis.rs [hunk 5]
```diff
@@ -0,0 +1,598 @@
+use aho_corasick::{AhoCorasick, AhoCorasickBuilder, MatchKind};
+use once_cell::sync::Lazy;
+
+/// Returns a redis command if it is a case-insensitive prefix of `seek`.
+pub fn matching_redis_command(seek: &str) -> Option<&str> {
+    let m = REDIS_COMMAND_MATCHER.find(&format!("{seek} "))?;
```
Comment (jjbayer) [inline relay-event-normalization/src/normalize/span/description/redis.rs]: The only suboptimal part of this solution is that it adds a string allocation for every search string.

**Comment by jjbayer** [comment]
@Dav1dde restored my Aho-Corasick approach now. Only thing I don't like is the extra string allocation.

File: relay-event-normalization/src/normalize/span/description/mod.rs [hunk 6]
```diff
@@ -384,15 +386,14 @@ pub fn concatenate_host_and_port(host: Option<&str>, port: Option<u16>) -> Cow<s
...
-    let scrubbed = match parts {
-        Some((Some(command), Some(_args))) => command.as_str().to_owned() + " *",
-        Some((Some(command), None)) => command.as_str().into(),
-        None | Some((None, _)) => "*".into(),
-    };
-    Some(scrubbed)
+    let string = string.trim();
+    Some(match matching_redis_command(string) {
+        Some(command) => match string.get(command.len()..) {
+            None | Some("") => command.to_string(),
```
Comment (Dav1dde) [inline relay-event-normalization/src/normalize/span/description/mod.rs]:
```suggestion
            None | Some("") => command.to_owned(),
```

File: relay-event-normalization/src/normalize/span/description/mod.rs [hunk 7]
```diff
@@ -384,15 +386,14 @@ pub fn concatenate_host_and_port(host: Option<&str>, port: Option<u16>) -> Cow<s
...
+            Some(_other) => format!("{command} *"),
+        },
+        None => "*".to_string(),
```
Comment (Dav1dde) [inline relay-event-normalization/src/normalize/span/description/mod.rs]:
```suggestion
        None => "*".to_owned(),
```

File: relay-event-normalization/src/normalize/span/description/redis.rs [hunk 8]
```diff
@@ -0,0 +1,598 @@
+use aho_corasick::{AhoCorasick, AhoCorasickBuilder, MatchKind};
+use once_cell::sync::Lazy;
```
Comment (Dav1dde) [inline relay-event-normalization/src/normalize/span/description/redis.rs]: The std lib has a `LazyLock` now.

File: relay-event-normalization/src/normalize/span/description/redis.rs (hunk 5 above)
Comment (Dav1dde) [reply relay-event-normalization/src/normalize/span/description/redis.rs]: This should be possible with two matchers, one with spaces at the end the other one without, if the matcher without spaces matches (after the one without failed) it must be the full command with no trailing data.

Alternative option, you can `find_iter{overlapping}` to get all matches then just select the longest one which has a space or end of input following. That seems pretty straight forward.

File: relay-event-normalization/src/normalize/span/description/redis.rs [hunk 9]
```diff
@@ -0,0 +1,598 @@
...
+    "ZREVRANGEBYSCORE ",
+    "ZREVRANK ",
+    "ZSCAN ",
+    "ZSCORE ",
+    "ZUNION ",
+    "ZUNIONSTORE ",
+];
+
+static REDIS_COMMAND_MATCHER: Lazy<AhoCorasick> = Lazy::new(|| {
+    AhoCorasickBuilder::new()
```
Comment (Dav1dde) [inline relay-event-normalization/src/normalize/span/description/redis.rs]: I think you want to set this to an anchored search mode `start_kind(StartKind::Anchored)`?

File: relay-event-normalization/src/normalize/span/description/redis.rs [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
...OCH|CLUSTER SAVECONFIG|CLUSTER RESET|CLUSTER REPLICATE|CLUSTER REPLICAS|CLUSTER NODES|CLUSTER MYSHARDID|CLUSTER MYID|CLUSTER MEET|CLUSTER LINKS|CLUSTER KEYSLOT|CLUSTER INFO|CLUSTER HELP|CLUSTER GETKEYSINSLOT|CLUSTER FORGET|CLUSTER FLUSHSLOTS|CLUSTER FAILOVER|CLUSTER DELSLOTSRANGE|CLUSTER DELSLOTS|CLUSTER COUNTKEYSINSLOT|CLUSTER COUNT-FAILURE-REPORTS|CLUSTER BUMPEPOCH|CLUSTER ADDSLOTSRANGE|CLUSTER ADDSLOTS|CLUSTER|CLIENT UNPAUSE|CLIENT UNBLOCK|CLIENT TRACKINGINFO|CLIENT TRACKING|CLIENT SETNAME|CLIENT SETINFO|CLIENT REPLY|CLIENT PAUSE|CLIENT NO-TOUCH|CLIENT NO-EVICT|CLIENT LIST|CLIENT KILL|CLIENT INFO|CLIENT ID|CLIENT HELP|CLIENT GETREDIR|CLIENT GETNAME|CLIENT CACHING|CLIENT|CF.SCANDUMP|CF.RESERVE|CF.MEXISTS|CF.LOADCHUNK|CF.INSERTNX|CF.INSERT|CF.INFO|CF.EXISTS|CF.DEL|CF.COUNT|CF.ADDNX|CF.ADD|BZPOPMIN|BZPOPMAX|BZMPOP|BRPOPLPUSH|BRPOP|BLPOP|BLMPOP|BLMOVE|BITPOS|BITOP|BITFIELD_RO|BITFIELD|BITCOUNT|BGSAVE|BGREWRITEAOF|BF.SCANDUMP|BF.RESERVE|BF.MEXISTS|BF.MADD|BF.LOADCHUNK|BF.INSERT|BF.INFO|BF.EXISTS|BF.CARD|BF.ADD|AUTH|ASKING|APPEND|ACL WHOAMI|ACL USERS|ACL SETUSER|ACL SAVE|ACL LOG|ACL LOAD|ACL LIST|ACL HELP|ACL GETUSER|ACL GENPASS|ACL DRYRUN|ACL DELUSER|ACL CAT|ACL)(?-u:\s|$)").unwrap()
```
Comment (Dav1dde) [inline relay-event-normalization/src/normalize/span/description/redis.rs]: Should we still have the list of commands as a spearate `const` and just build the regex once with a loop?

Or maybe format differently, seems hard to parse, but nbd.

Comment (Dav1dde) [inline relay-event-normalization/src/normalize/span/description/redis.rs]: I think you can disable unicode for the regex with `(?-u)` in the beginning, that's what [globset does](https://github.com/BurntSushi/ripgrep/blob/master/crates/globset/src/glob.rs#L641-L656).

File: Cargo.toml [hunk 11]
```diff
@@ -61,6 +61,7 @@ relay-protocol-derive = { path = "relay-protocol-derive" }
 relay-event-derive = { path = "relay-event-derive" }
 
 ahash = "0.8.11"
+aho-corasick = "1.1.3"
```
Comment (Dav1dde) [inline Cargo.toml]:
```suggestion
```

Leftover

File: relay-event-normalization/Cargo.toml [hunk 12]
```diff
@@ -13,6 +13,7 @@ publish = false
 workspace = true
 
 [dependencies]
+aho-corasick = { workspace = true }
```
Comment (Dav1dde) [inline relay-event-normalization/Cargo.toml]:
```suggestion
```

Leftover

**Review by Dav1dde — APPROVED** [review]

File: Cargo.toml (hunk 11 above)
Comment (jjbayer) [reply Cargo.toml]: Thanks.

File: relay-event-normalization/src/normalize/span/description/redis.rs (hunk 10 above)
Comment (jjbayer) [reply relay-event-normalization/src/normalize/span/description/redis.rs]: It complained because I did not escape dots in redis commands. Fixed now.
